compiler/src/dotty/tools/dotc/typer/Inferencing.scala

```diff
@@ -635,14 +635,27 @@ trait Inferencing { this: Typer =>
                 // isInstantiated needs to be checked again, since previous interpolations could already have
                 // instantiated `tvar` through unification.
                 val v = vs(tvar)
+                val tparam = tvar.origin
                 if v == null then buf += ((tvar, 0))
-                else if v.intValue != 0 then buf += ((tvar, v.intValue))
+                else if v.intValue != 0 then
+                  def hasLowerBound =
+                    (constraint.nonParamBounds(tparam).lo ne tparam.underlying.bounds.lo)
+                    || constraint.lower(tparam).nonEmpty
+                  def avoidNothing =
+                    (tp eq tvar) && pt.isInstanceOf[SelectionProto]
+                  if v.intValue != 1 || hasLowerBound || !avoidNothing then
```

Comment:
I would maybe call this "isSelectionPrefix" since the lower-bound isn't necessarily Nothing (in particular it might be Null)
```suggestion
                  def isSelectionPrefix =
                    (tp eq tvar) && pt.isInstanceOf[SelectionProto]
                  if v.intValue != 1 || hasLowerBound || !isSelectionPrefix then
```